scripts/main/album.js

```diff
@@ -937,6 +937,131 @@ album.setPublic = function (albumID, e) {
 	});
 };
 
+album.shareUsers = function (albumID, e) {
+	var addShare = function (listIds) {
+		var params = {
+			albumIDs: albumID,
+			UserIDs: listIds.join(","),
+		};
+		api.post("Sharing::Add", params, (data) => {
+			if (data !== true) {
+				loadingBar.show("error", data.description);
+				lychee.error(null, params, data);
+			} else {
+				loadingBar.show("success", "Sharing updated!");
+			}
+		});
+	};
+
+	var removeShare = function (listIds) {
+		var params = { ShareIDs: listIds.join(",") };
+		api.post("Sharing::Delete", params, function (data) {
+			if (data !== true) {
+				loadingBar.show("error", data.description);
+				lychee.error(null, params, data);
+			}
+		});
+	};
+
+	if (!basicModal.visible()) {
+		let msg = `<form id="sharing_people_form">
+			<p>${lychee.locale["WAIT_FETCH_DATA"]}</p>
+		</form>`;
+
+		api.post("User::List", {}, (users) => {
+			$("#sharing_people_form").empty();
+			if (users !== {}) {
+				$("#sharing_people_form").append(`<p>${lychee.locale["SHARING_ALBUM_USERS_LONG_MESSAGE"]}</p>`);
+				$.each(users, (_, user) => {
+					$("#sharing_people_form").append(`<div class='choice'>
+						<label>
+							<input type='checkbox' name='${user.id}'>
+							<span class='checkbox'>${build.iconic("check")}</span>
+							<span class='label'>${user.username}</span>
+						</label>
+						<p></p>
+					</div>`);
+				});
+				api.post("Sharing::List", {}, (data) => {
+					if (data !== undefined && data.shared !== []) {
+						var sharingOfAlbum = data.shared.filter((val) => val.album_id == albumID);
+						sharingOfAlbum.forEach((sharing) => {
+							$(`.basicModal .choice input[name="${sharing.user_id}"]`).prop("checked", true);
+						});
+					}
+				});
+			} else {
+				$("#sharing_people_form").append(`<p>${lychee.locale["SHARING_ALBUM_USERS_NO_USERS"]}</p>`);
+			}
+		});
+
+		basicModal.show({
+			body: msg,
+			buttons: {
+				action: {
+					title: lychee.locale["ALBUM_SHARING_CONFIRM"],
+					fn: (data) => {
+						album.shareUsers(albumID, e);
+					},
+				},
+				cancel: {
+					title: lychee.locale["CANCEL"],
+					fn: basicModal.close,
+				},
+			},
+		});
+		return true;
+	}
+
+	var usersToShare = [];
+	var usersNotToShare = [];
+	$(".basicModal .choice input").each((_, input) => {
+		if ($(input).is(":checked")) {
+			usersToShare.push(parseInt(input.name, 10));
+		} else {
+			usersNotToShare.push(parseInt(input.name, 10));
+		}
+	});
+
+	api.post("Sharing::List", {}, (data) => {
```

Comment:
Didn't we invoke it earlier when opening the modal? Couldn't we cache the response from then? All you need to know now are the initial states and sharing `id`'s, correct? Couldn't you store them, say, in a `data-` attribute of each checkbox input?

Reply:
> All you need to know now are the initial states and sharing id's, correct?

Exactly. I did the API call twice, because I didn't figure out a way to keep the result of the call to `Sharing::List` when processing the choices of the user.

> Couldn't you store them, say, in a data- attribute of each checkbox input?

Yes I think it would work.